staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -687,13 +783,36 @@ func (c *Cacher) dispatchEvents() {
 
 func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 	c.startDispatching(event)
+	defer c.finishDispatching()
+	// Watchers stopped after startDispatching will be delayed to finishDispatching,
 
 	// Since add() can block, we explicitly add when cacher is unlocked.
-	for _, watcher := range c.watchersBuffer {
-		watcher.add(event, c.timer, c.dispatchTimeoutBudget)
+	if event.Type == watch.Bookmark {
+		for _, watcher := range c.watchersBuffer {
+			watcher.nonblockingAdd(event)
```

Comment:
Do we need to check return value here ?

Reply:
Ignored intentionally, bookmark is not necessary when watch is blocked(which means there are many events recently).